File: PreCourse/Merging and Searching/rotationCount.cpp
#include <bits/stdc++.h>
using namespace std;

int rotationCount(int a[], int size)
{
    for (int i = 0; i < size - 1; ++i)
    {
        if (a[i] > a[i + 1])
        {
            return i + 1;
        }
    }
    return 0;
}

int main()
{
    int T, i, size = 0;
    cin >> T;

    while (T--)
    {
        cin >> size;

        int arr[size];
        for (i = 0; i < size; i++)
            cin >> arr[i];

        cout << rotationCount(arr, size) << endl;
    }
    return 0;
}

File: SuperCoders/Week 3/Day 5/RotationCount.java
class Result {
    int rotationCount(int arr[], int size) {
        if (!(arr[0] > arr[size - 1])) {
            return 0;
        }

        int start = 0;
        int end = size - 1;
        while (start <= end) {
            int mid = (start + end) / 2;
            if (mid == 0) {
                return arr[mid] > arr[mid + 1] ? 1 : 0;
            }
            else if (mid == size - 1) {
                return arr[mid] < arr[mid - 1] ? mid : 0;
            }
            else {
                if (arr[mid] < arr[mid - 1] && arr[mid] < arr[mid + 1]) {
                    return mid;
                }
                else if (arr[start] > arr[mid]) {
                    end = mid;
                }
                else {
                    start = mid;
                }
            }
        }
        return 0;
    }
}